### ISSUE TITLE
Unhandled exception in call to Workspace.TryApplyChanges

### ISSUE THREAD

**Issue by rik-smeets** [dotnet/format#1623]
The method `TryApplyChanges` on `Workspace` can throw a `NotSupportedException`. This can for example occur due to the issue described [here](https://github.com/dotnet/roslyn/issues/55014#issuecomment-1141222922).

The method call in question from this solution:
https://github.com/dotnet/format/blob/01ea81e043c2b4b24fb67fc8c73a95d1e4d3fb6b/src/CodeFormatter.cs#L101

The `NotSupportedException` should be handled here. Right now, the format action simply crashes without much information to go on (even with verbose logging enabled):
```
C:\Git\DotNetFormatTest>dotnet format
Unhandled exception: System.NotSupportedException: Changing document properties is not supported
   at Microsoft.CodeAnalysis.Workspace.CheckAllowedProjectChanges(ProjectChanges projectChanges)
   at Microsoft.CodeAnalysis.Workspace.CheckAllowedSolutionChanges(SolutionChanges solutionChanges)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution)
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.FormatWorkspaceAsync(FormatOptions formatOptions, ILogger logger, CancellationToken cancellationToken, String binaryLogPath)
   at Microsoft.CodeAnalysis.Tools.FormatCommandCommon.FormatAsync(FormatOptions formatOptions, ILogger`1 logger, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Tools.Commands.RootFormatCommand.FormatCommandDefaultHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

**Comment by gruber76** [comment]
I, too, am seeing this issue attempting to run `dotnet format` with dotnet CLI version 6.0.301

**Comment by Neogentai** [comment]
I am also running into this issue.

**Comment by sprudel79** [comment]
Same issue for me with

```
dotnet format --version
6.4.345109+d2f060021d5a71871927bb961a6c09e76bad2be7
```

**Comment by warappa** [comment]
Also happens with 7.x version.

### Test case
If `IDE0130` and `dotnet_style_namespace_match_folder` are both set to `warning` and a wrong namespace exists in code, then the mentioned exception happens.

```ini
dotnet_style_namespace_match_folder = true:warning
dotnet_diagnostic.IDE0130.severity = warning
```

This example error seems like this ticket (closed as fixed):
https://github.com/dotnet/roslyn/issues/54757

### Info
dotnet format --version
7.0.352001+6a22157622e5da71ba0c43d23269352210bdb277

dotnet --version
7.0.100

dotnet --info

```log
.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  5.0.414 [C:\Program Files\dotnet\sdk]
  7.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.29 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  C:\src\global.json
```

**Comment by avidenic** [comment]
Seems to be fixed with 7.0.101

**Comment by mr-davidc** [comment]
This happens on version `7.0.102` for me.

dotnet format --version
`7.0.357101+49c2ef651359526841d13e66129b71d1bcd9cef9`

dotnet --version
`7.0.102`

dotnet --info

```
.NET SDK:
 Version:   7.0.102
 Commit:    4bbdd14480

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.102\

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  3.1.404 [C:\Program Files\dotnet\sdk]
  5.0.416 [C:\Program Files\dotnet\sdk]
  7.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by idilov** [comment]
Still happens on sdk 7.0.203. This
```ini
dotnet_diagnostic.IDE0130.severity = warning
```
is enough to crash `dotnet format`. No need to also enable dotnet_style_namespace_match_folder.
Sorry to troll, but it's been a year since the OP so I decided to remind of this probably forgotten issue.

**Comment by savbace** [comment]
Reproduces for me as well.
SDK version: 7.0.402
```sh
dotnet format --version
7.4.437506+270bbfa0b86b9158575ba986dff9244ed4561192
```

**Comment by khdevnet** [comment]
> Reproduces for me as well. SDK version: 7.0.402
> 
> ```shell
> dotnet format --version
> 7.4.437506+270bbfa0b86b9158575ba986dff9244ed4561192
> ```

Have the same issue in the version  7.4.431902+3c30490bb2ecae8967a3b5e79c97ab98de334676

**Comment by manzur-billiyo** [comment]
Reproduces for me as well on .NET 8

`dotnet format --version
8.0.453106+2651752953c0d41c8c7b8d661cf2237151af33d0
`

**Comment by lofcz** [comment]
Reproduced on `8.0.453106+2651752953c0d41c8c7b8d661cf2237151af33d0`

**Comment by mikesigs** [comment]
And still reproducable on 8.0.100

**Comment by westthomasgoodenergy** [comment]
I too am seeing this on dotnet 8.

dotnet format --version
8.0.453106+2651752953c0d41c8c7b8d661cf2237151af33d0

**Comment by wtorricos** [comment]
I'm seeing it as well.
dotnet format --version
8.0.511601+d237e172b324021b97effa244af44d63d1a8bb7e

**Comment by sharwell** [comment]
The short-term resolution to this issue is to update dotnet format to report a warning if IDE0130 appears, and simply ignore any attempt to fix it. The code fix for IDE0130 is not compatible with MSBuildWorkspace that dotnet format uses, so dotnet format will not be able to apply these fixes automatically.

**Comment by NikiforovAll** [comment]
Any updates on this one? I have an idea to bring "Sync Namespaces" right-click to VS Code. But this issue doesn't allow me to apply the fix properly.

Repo: https://github.com/NikiforovAll/vsc-sync-namespaces-dotnet

**Comment by jgeurts** [comment]
Adding the following to .editorconfig fixed it for me:

```
dotnet_diagnostic.IDE0130.severity = suggestion
```

**Comment by snerte** [comment]
Having the same issue in `9.0.100-rtm.24529.9+59db016f11bb27d359336cf37524b863d77e7fea`, two years after this issue was reported 😶‍🌫️

Adding any of these lines causes the .net format to crash:

``` 
dotnet_style_namespace_match_folder = true:warning
dotnet_diagnostic.IDE0130.severity = warning
```

> TLDR for those who just want to get on with their work: use the `--exclude-diagnostics IDE0130` flag. This makes the final command `dotnet format --exclude-diagnostics IDE0130`.

Omits the issue for us 💘

> Adding the following to . editorconfig fixed it for me:
> 
> ```
> dotnet_diagnostic.IDE0130.severity = suggestion
> ```

Setting it to `suggestion` is no option, since we're using the `/p:TreatWarningsAsErrors=true` flag to treat warnings as errors in our pipelines, but want to show warnings to developers while devving.
